calm_adapter/calm_deletion_checker/src/main/scala/uk/ac/wellcome/platform/calm_deletion_checker/DeletionChecker.scala

```diff
@@ -1,14 +1,81 @@
 package uk.ac.wellcome.platform.calm_deletion_checker
 
-import uk.ac.wellcome.platform.calm_api_client.CalmApiClient
+import uk.ac.wellcome.platform.calm_api_client.{
+  CalmApiClient,
+  CalmQuery,
+  CalmSession
+}
 import weco.catalogue.source_model.CalmSourcePayload
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
+
+trait DeletionChecker {
+  type Record = CalmSourcePayload
+  type Records = Set[Record]
+
+  protected def nDeleted(records: Records): Future[Int]
+
+  // The implementation here is based on the algorithm given in Wang et al (2017)
+  // Found at: https://arxiv.org/abs/1407.2283, see `Algorithm 1`
+  //
+  // While there are more optimal algorithms that exist, the implementation of this one is very simple
+  def deletedRecords(batch: Records)(
+    implicit ec: ExecutionContext): Future[Records] = {
+    def findDeleted(records: Records, d: Int): Future[Records] = d match {
+      case 0                      => Future.successful(Set.empty)
+      case n if n == records.size => Future.successful(records)
+      case _ =>
+        val testSet = records.take(M(records.size, d))
+        for {
+          d1 <- nDeleted(testSet)
+          deletions1 <- findDeleted(testSet, d1)
+          deletions2 <- findDeleted(records -- testSet, d - d1)
+        } yield deletions1 ++ deletions2
+    }
+
+    nDeleted(batch).flatMap { d =>
+      findDeleted(batch, d)
+    }
+  }
+
+  // See equations (6), (7) in https://arxiv.org/abs/1407.2283
+  private def l(n: Int, d: Int): Int = math.ceil(log2(n.toFloat / d)).toInt - 1
+  private def k(n: Int, d: Int): Int =
+    math.ceil(n / math.pow(2, l(n, d))).toInt - d
+
+  // See equation (11)
+  private def M(n: Int, d: Int): Int =
+    if (d <= n / 2) {
+      n - math.pow(2, l(n, d)).toInt * (d + k(n, d) - 1)
+    } else M(n, n - d)
+
+  // See equation (5) in https://arxiv.org/abs/1407.2283
+  // To account for the initial count (outside of the algorithm proper)
+  // we remove the (-1) term and ensure a minimum of 1 test.
+  //
+  // This is only used for testing the implementation
+  def nTestsUpperBound(n: Int, d: Int): Int =
+    math.max((l(n, d) + 1) * d + k(n, d), 1)
+
+  private def log2(x: Float): Float = (math.log(x) / math.log(2)).toFloat
+}
 
-class DeletionChecker(calmApiClient: CalmApiClient) {
+class ApiDeletionChecker(calmApiClient: CalmApiClient)(
+  implicit ec: ExecutionContext)
+    extends DeletionChecker {
 
-  def deletedRecords(
-    batch: Seq[CalmSourcePayload]): Future[Set[CalmSourcePayload]] =
-    ???
+  def nDeleted(records: Records): Future[Int] =
+    calmApiClient
```

Comment:
Will how Calm deals with this query affect the performance of this implementation? If for example the time taken to return a result scales linearly with the number of ids to search for? I have not yet had any coffee, so please excuse me if this is a stupid question.

Reply:
It is not a stupid question - yes, it will affect the performance, but while I didn't benchmark anything properly when I was testing out the big queries they seemed to scale sub-linearly (or at least the things it did that might have scaled linearly were dwarfed by other constant factors). There is probably an optimal batch size to balance API perf and number of queries but for now I think we'll just see if this is good enough...